src/boot/efi/linux.c

```diff
@@ -13,83 +13,22 @@
 
 #include "initrd.h"
 #include "linux.h"
+#include "missing_efi.h"
 #include "pe.h"
 #include "util.h"
 
-static EFI_LOADED_IMAGE_PROTOCOL *loaded_image_free(EFI_LOADED_IMAGE_PROTOCOL *img) {
-        if (!img)
-                return NULL;
-        mfree(img->LoadOptions);
-        return mfree(img);
-}
-
-static EFI_STATUS loaded_image_register(
-                const char *cmdline, UINTN cmdline_len,
-                const void *linux_buffer, UINTN linux_length,
-                EFI_HANDLE *ret_image) {
-
-        EFI_LOADED_IMAGE_PROTOCOL *loaded_image = NULL;
-        EFI_STATUS err;
-
-        assert(cmdline || cmdline_len > 0);
-        assert(linux_buffer && linux_length > 0);
-        assert(ret_image);
-
-        /* create and install new LoadedImage Protocol */
-        loaded_image = xnew(EFI_LOADED_IMAGE_PROTOCOL, 1);
-        *loaded_image = (EFI_LOADED_IMAGE_PROTOCOL) {
-                .ImageBase = (void *) linux_buffer,
-                .ImageSize = linux_length
-        };
-
-        /* if a cmdline is set convert it to UCS2 */
-        if (cmdline) {
-                loaded_image->LoadOptions = xstra_to_str(cmdline);
-                loaded_image->LoadOptionsSize = strsize16(loaded_image->LoadOptions);
-        }
-
-        /* install a new LoadedImage protocol. ret_handle is a new image handle */
-        err = BS->InstallMultipleProtocolInterfaces(
-                        ret_image,
-                        &LoadedImageProtocol, loaded_image,
-                        NULL);
-        if (err != EFI_SUCCESS)
-                loaded_image = loaded_image_free(loaded_image);
-
-        return err;
-}
-
-static EFI_STATUS loaded_image_unregister(EFI_HANDLE loaded_image_handle) {
-        EFI_LOADED_IMAGE_PROTOCOL *loaded_image;
+EFI_STATUS load_trusted_image(EFI_HANDLE parent, const void *source, size_t len, EFI_HANDLE *ret_image) {
```

Comment:
"trusted" sounds as the image was already trusted here, but that's not the case, the firmware will do its authentication, and only *then* it is trusted... So right here it is in a netherworld where we trust it (because it's already covered by our SecureBoot signatre) but the firmware doesn't. 

Maybe "load_image_and_authenticate()"?